clang/lib/CIR/Lowering/ThroughMLIR/LowerCIRToMLIR.cpp

```diff
@@ -208,6 +209,18 @@ class CIRCosOpLowering : public mlir::OpConversionPattern<mlir::cir::CosOp> {
   }
 };
 
+class CIRSinOpLowering : public mlir::OpConversionPattern<mlir::cir::SinOp> {
```

Comment:
May I suggest to introduce a template for the 1:1 lowerings from `cir` to `math` ops instead? Here, it would just be used by the existing `cos` and new `sin` pattern, but looking at #592, the boilerplate duplication would become quite significant otherwise.

Reply:
Yes, you are right, we can use a template for the 1:1 lowerings from `cir` to mlir `math` ops, and we can also use the same template for the 1:1 lowerings from `cir` to llvm `math` ops. But the `cir.rint` `cir.shift` ops cannot use template, these ops should use the old way.

Reply:
Instead of using templates I'd rather leverage tablegen to do these. There's a similar idea introduced for LLVM lowering in https://github.com/llvm/clangir/pull/434 we could leverage, once that lands (I might commander that PR soon, but happy to let someone else take it) we could do similar for MLIR lowering.